### PR TITLE
Test that connections that should be permutations actually *are* permutations

### PR DESCRIPTION
**Pull request by inducer**
This is supposed to be a version of #105 that we can merge right away, plus a number of fixes that arose around that. For review, this PR is probably best viewed commit-by-commit. The commit messages provide a pretty good inventory of what's here.

Diff-wise, the biggest part of this is the (annoying, but IMO required) split of `PolynomialWarpAndBlendGroupFactory` into `PolynomialWarpAndBlend2DGroupFactory` and `PolynomialWarpAndBlend3DGroupFactory`. (The 2D WnB nodes are *not* equivalent to a restriction of the 3D nodes to a face, and we very much would like 2D el groups in 3D discretizations to be just that, so that the boundary extractions can be done by indirect addressing.)

- [ ] Post-merge, this needs a corresponding (and hopefully much smaller) change in grudge.

cc @nchristensen

### CODE REVIEW

File: meshmode/discretization/connection/direct.py [hunk 1]
```diff
@@ -250,28 +250,36 @@ def _resample_point_pick_indices(self, actx: ArrayContext,
...
-                return None
-            if len(zero_indices) != ncols - 1:
+        dist_vecs = (ibatch.result_unit_nodes.reshape(dim, -1, 1)
+                - from_grp.unit_nodes.reshape(dim, 1, -1))
+        dists = np.sqrt(np.sum(dist_vecs**2, axis=0))
+
+        result = np.zeros(ntgt_nodes, dtype=self.to_discr.mesh.element_id_dtype)
+
+        if tol_multiplier is None:
+            tol_multiplier = 500
```
Comment (alexfikl) [inline meshmode/discretization/connection/direct.py:273]: There's another `tol_multiplier is None` above that sets it to 50. Which one did you mean to keep?

File: meshmode/discretization/connection/direct.py [hunk 2]
```diff
@@ -250,28 +250,36 @@ def _resample_point_pick_indices(self, actx: ArrayContext,
...
+        if dim == 0:
+            assert ntgt_nodes == 1
+            return actx.freeze(actx.from_numpy(np.array([0], dtype=np.int32)))
 
-            if len(one_indices) != 1:
-                return None
-            if len(zero_indices) != ncols - 1:
+        dist_vecs = (ibatch.result_unit_nodes.reshape(dim, -1, 1)
+                - from_grp.unit_nodes.reshape(dim, 1, -1))
+        dists = np.sqrt(np.sum(dist_vecs**2, axis=0))
```
Comment (alexfikl) [inline meshmode/discretization/connection/direct.py:268]:
```suggestion
        dists = la.norm(dist_vecs, axis=0, ord=2)
```
? Don't know if this is any faster.

File: meshmode/discretization/poly_element.py [hunk 3]
```diff
@@ -287,6 +289,41 @@ class PolynomialWarpAndBlendElementGroup(_MassMatrixQuadratureElementGroup):
     interpolation. Interpolation nodes edge-clustered for avoidance of Runge
     phenomena. Nodes are present on the boundary of the simplex.
 
+    Uses :func:`modepy.warp_and_blend_nodes`.
+    """
+    def __init__(self, mesh_el_group, order, index):
```
Comment (alexfikl) [inline meshmode/discretization/poly_element.py:294]: I'm not sure how I feel about this splitting of the warp and blend classes.

A couple of random thoughts:
* If I'm understanding the issue correctly, this can also be solved by just introducing the 3D version and being careful to pass it to `make_face_restriction`? Is that too much to leave to the user to get right?
* How about a flag in the original `PolynomialWarpAndBlendElementGroup` like `restricted_from_3d=None` (that gets set to `True` or `False` based on dimension)?

File: meshmode/discretization/poly_element.py [hunk 4]
```diff
@@ -304,6 +344,35 @@ def _interp_nodes(self):
         return result
 
 
+class PolynomialWarpAndBlend3DElementGroup(_MassMatrixQuadratureElementGroup):
```
Comment (alexfikl) [inline meshmode/discretization/poly_element.py:347]: Feels like this should have `From3DRestriction` or something to that effect in the name? This makes it sound like it only works in 3D.

File: meshmode/interop/firedrake/connection.py [hunk 5]
```diff
@@ -39,9 +39,8 @@
 from meshmode.mesh.processing import get_simplex_element_flip_matrix
 
 from meshmode.discretization.poly_element import (
-    PolynomialWarpAndBlendGroupFactory,
-    PolynomialRecursiveNodesGroupFactory,
-    ElementGroupFactory)
+        default_simplex_group_factory,
+        ElementGroupFactory)
```
Comment (alexfikl) [inline meshmode/interop/firedrake/connection.py:43]:
```suggestion
    default_simplex_group_factory,
    ElementGroupFactory)
```
? Everyone else seems to be indented like this.

File: meshmode/discretization/poly_element.py [hunk 6]
```diff
@@ -304,6 +344,35 @@ def _interp_nodes(self):
...
+    """
+    @property
+    @memoize_method
+    def _interp_nodes(self):
+        dim = self.mesh_el_group.dim
+        if self.order == 0:
+            result = mp.warp_and_blend_nodes(dim, 1)
+            result = np.mean(result, axis=1).reshape(-1, 1)
+        else:
+            result = mp.warp_and_blend_nodes(3, self.order)
```
Comment (alexfikl) [inline meshmode/discretization/poly_element.py:365]: How nice are the restrictions of the 3D nodes? In terms of Lebesgue constants and things like that.

**Review by alexfikl — APPROVED** [review]
Not quite sure about the split of the warp and blend classes, but functionally it looks good to me! (barring a few nitpicks)

File: meshmode/discretization/poly_element.py (hunk 4 above)
Comment (inducer) [reply meshmode/discretization/poly_element.py:347]: Fair point.

* `PolynomialWarpAndBlend3DRestrictingElementGroup`?
* `PolynomialWarpAndBlend2DRestrictingElementGroup`?

Comment (alexfikl) [reply meshmode/discretization/poly_element.py:347]: Yep, that sounds better!

File: meshmode/discretization/poly_element.py (hunk 3 above)
Comment (inducer) [reply meshmode/discretization/poly_element.py:294]: > I'm not sure how I feel about this splitting of the warp and blend classes.

I can see where you're coming from. But to me they simply are two mostly unrelated sets of nodes. If we distinguished them by a flag in `PolynomialWarpAndBlendElementGroup`, it would still be like two functions wedged into one with an `if` statement.

And the original `PolynomialWarpAndBlendElementGroup` is just not worth having: For DG, it'll automatically be inefficient.

> Is that too much to leave to the user to get right?

If you'd like to add a special `if` statement for this into `make_face_restriction`, maybe not. Otherwise, I think it's hard to get right: It's totally legitimate to ask for any old node set you like, and if we don't do *something* like the split I'm proposing, you almost automatically end up with 3D->2D nodes that unnecessarily require interpolation.

> How about a flag in the original PolynomialWarpAndBlendElementGroup like restricted_from_3d=None (that gets set to True or False based on dimension)?

To my mind, the issue is that the factory has no way of telling that a 2D group is descended from a 3D group when it creates it. We could set state in the factory based on the first group created, but find that too ugly/problematic to seriously consider.

File: meshmode/discretization/poly_element.py (hunk 6 above)
Comment (inducer) [reply meshmode/discretization/poly_element.py:365]: https://gist.github.com/inducer/bbe9fd6e32d5e67b3d4e446479cfa633

tl;dr: Negligibly different IMO.

File: meshmode/discretization/poly_element.py (hunk 3 above)
Comment (alexfikl) [reply meshmode/discretization/poly_element.py:294]: Yeah, I'm pretty sold on the fact that they're basically two different family of nodes and they both have to be there.

With the better naming, I'm happy to close this random rant!

File: meshmode/discretization/poly_element.py (hunk 6 above)
Comment (alexfikl) [reply meshmode/discretization/poly_element.py:365]: Nice! :tada:

File: meshmode/interop/firedrake/connection.py (hunk 5 above)
Comment (inducer) [reply meshmode/interop/firedrake/connection.py]: Gone in 8a822fe.

File: meshmode/discretization/poly_element.py (hunk 4 above)
Comment (inducer) [reply meshmode/discretization/poly_element.py]: 8a822fe

File: meshmode/discretization/connection/direct.py (hunk 2 above)
Comment (inducer) [reply meshmode/discretization/connection/direct.py]: ba86136

File: meshmode/discretization/connection/direct.py (hunk 1 above)
Comment (inducer) [reply meshmode/discretization/connection/direct.py]: ba86136
